### PR TITLE
Fix/pallas tests shared memory

### PR DESCRIPTION
**Pull request by AratiGanesh**
## Motivation

Pallas GPU tests (gpu_ops_test.py and ops_test.py) were crashing with RESOURCE_EXHAUSTED errors on devices (MI250) with limited shared memory. This PR aims to estimate the shared memory and ensure that tests either adjust to fit the available memory or skip if they cannot run.

## Technical Details
1. gpu_ops_test.py - Uses similar logic as https://github.com/ROCm/jax/pull/559/files.
Implements a shared memory estimate function for fused attention.  Tests now automatically reduce block sizes (128x128 → 32x32) if the original config exceeds device limits. The backward pass logic also scales its gradients blocks proportionally to the forward pass.

2. ops_test.py - Estimates the share memory and if it cannot fit matrices, it skips the test. 

## Test Plan

Rerun 
    pytest tests/pallas/gpu_ops_test.py -v 
    pytest jax/tests/pallas/gpu_ops_test.py -v

## Test Result

All tests pass. 

## Submission Checklist

- [ ] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

### CODE REVIEW

File: build/rocm-test-requirements.txt
Comment (mminutoli) [inline build/rocm-test-requirements.txt:1]: Can we take this out of this PR?

I'd like to be an "atomic" change fixing only 1 issue that we can send to upstream jax.

File: tests/pallas/gpu_ops_test.py [hunk 1]
```diff
@@ -164,17 +218,32 @@ def setUp(self):
...
-      *,
-      batch_size,
-      seq_len,
-      num_heads,
-      head_dim,
-      block_sizes,
-      causal,
-      use_fwd,
-      use_segment_ids,
-  ):
```
Comment (mminutoli) [inline tests/pallas/gpu_ops_test.py:177]: Let's not change their formatting.

File: tests/pallas/gpu_ops_test.py [hunk 2]
```diff
@@ -164,17 +218,32 @@ def setUp(self):
...
+          use_fwd,
+          use_segment_ids,
+      ):
+    # SHMEM Adjustment Logic
+    original_blocks = dict(block_sizes)
+    block_q_orig = original_blocks["block_q"]
+    block_k_orig = original_blocks["block_k"]
+    dtype = jnp.float16
+
+    adjusted_q, adjusted_k = self._adjust_mha_params_for_shared_memory(
```
Comment (mminutoli) [inline tests/pallas/gpu_ops_test.py:238]: I am ok with this like this, but I am wondering if they'll want this done only for GPUs.

**Review by mminutoli — CHANGES_REQUESTED** [review]
it's really minor comments with the intent of making it so it is easier to review from upstream. Please, fix them and let's create a PR also for upstream jax.
